### DISCUSSION TITLE
Image classification/statistics help

### DISCUSSION THREAD

**Discussion by olNick** [ImageMagick/ImageMagick#2271] [Help]
Hi all,
I have 8-bit greyscale PNG images and would like count the pixels from lets say 8 distinct ranges (possibly min-max)
This would allow me to the "classify" the as pixel count of the class / total count, resulting in an area percentage of each class.

I would also like to build a legend base on the classes like "Very low"------"Average"-------"very high" etc.
![EL434_env](https://user-images.githubusercontent.com/28807404/87549629-b9191e00-c6b6-11ea-9207-428d63de079a.png)

I'm new to IM and have tried the different histogram options, but I'm stuck
regards,
Nick

**Comment by snibgo** [comment]
"-posterize" will reduce the image to N shades of gray. Then a histogram gives the pixel count at each shade. For example, with a grayscale input:
```
magick in.png -posterize 8 -verbose info:
```
[EDIT: changed from "-posterize gray", which was wrong.]

The text output includes:
```
  Colors: 8
  Histogram:
       293: (    0,    0,    0) #000000000000 gray(0)
      3999: ( 9362, 9362, 9362) #249224922492 gray(14.2857%)
      7286: (18724,18724,18724) #492449244924 gray(28.5714%)
     23619: (28086,28086,28086) #6DB66DB66DB6 gray(42.8571%)
     17540: (37449,37449,37449) #924992499249 gray(57.1429%)
      5110: (46811,46811,46811) #B6DBB6DBB6DB gray(71.4286%)
      4233: (56173,56173,56173) #DB6DDB6DDB6D gray(85.7143%)
       131: (65535,65535,65535) #FFFFFFFFFFFF gray(255)
```

**Reply by olNick** [reply]
Thanx,
Actually this worked for me;
magick in.png -posterize 8 -verbose info:

Does not recognize "gray"

**Reply by snibgo** [reply]
Sorry, my copy-paste mistake. I intended "-posterize 8", not "-posterize gray".

**Comment by snibgo** [comment]
One characterisic of "-posterize" is that the bins at each end cover half the range of the other bins. So a grayscale gradient, a smooth transition from black to white, gives:
```
magick -size 1000x1000 gradient: -posterize 8 -verbose info:

  Colors: 8
  Histogram:
     72000: (    0,    0,    0) #000000000000 gray(0)
    143000: ( 9362, 9362, 9362) #249224922492 gray(14.2857%)
    142000: (18724,18724,18724) #492449244924 gray(28.5714%)
    143000: (28086,28086,28086) #6DB66DB66DB6 gray(42.8571%)
    143000: (37449,37449,37449) #924992499249 gray(57.1429%)
    142000: (46811,46811,46811) #B6DBB6DBB6DB gray(71.4286%)
    143000: (56173,56173,56173) #DB6DDB6DDB6D gray(85.7143%)
     72000: (65535,65535,65535) #FFFFFFFFFFFF gray(255)
```

If we want every bin to cover the same range, we can do this:
```
magick -size 1000x1000 gradient: -fx "int(u*8)/7" -clamp -verbose info:

  Colors: 8
  Histogram:
    125000: (    0,    0,    0) #000000000000 black
    125000: ( 9362, 9362, 9362) #249224922492 srgb(14.2857%,14.2857%,14.2857%)
    125000: (18724,18724,18724) #492449244924 srgb(28.5714%,28.5714%,28.5714%)
    125000: (28086,28086,28086) #6DB66DB66DB6 srgb(42.8571%,42.8571%,42.8571%)
    125000: (37449,37449,37449) #924992499249 srgb(57.1429%,57.1429%,57.1429%)
    125000: (46811,46811,46811) #B6DBB6DBB6DB srgb(71.4286%,71.4286%,71.4286%)
    125000: (56173,56173,56173) #DB6DDB6DDB6D srgb(85.7143%,85.7143%,85.7143%)
    125000: (65535,65535,65535) #FFFFFFFFFFFF white
```

**Reply by olNick** [reply]
Hi,
Could you expand on this, I'm missing it.. 
Also, an example of the "-k" flag below would be helpfull also.

regards

**Comment by fmw42** [comment]
You could also classify the image into 8 groups of colors (shades of gray) using -kmeans.
